src/sentry/static/sentry/app/utils/bulkController.tsx

```diff
@@ -0,0 +1,134 @@
+  /**
+   * Ready to be rendered summary component showing how many items are selected,
+   * with buttons to select everything, cancel everything, etc...
+   */
+  bulkNotice: React.ReactNode;
+} & Pick<State, 'selectedIds' | 'isEverythingSelected'>;
+
+type State = {
+  /**
+   * Selected ids on the current page
+   */
+  selectedIds: string[];
+  /**
+   * Are all rows across all pages selected?
+   */
+  isEverythingSelected: boolean;
+};
+
+type Props = {
+  /**
+   * Array of ids on current page
+   */
+  pageIds: string[];
+  /**
+   * Number of all rows across all pages
+   */
+  allIdsCount: number;
+  /**
+   * Number of grid columns to stretch the selection summary (used in BulkNotice)
+   */
+  noticeColumns: number;
+  /**
+   * Children with render props
+   */
+  children: (props: RenderProps) => React.ReactNode;
+  /**
+   * Maximum number of rows that can be bulk manipulated at once (used in BulkNotice)
+   */
+  maxIdsCount?: number;
+};
+
+class BulkController extends React.Component<Props, State> {
+  state: State = {
+    selectedIds: [],
+    isEverythingSelected: false,
+  };
+
+  static getDerivedStateFromProps(props: Props, state: State) {
+    return {
+      ...state,
+      selectedIds: intersection(state.selectedIds, props.pageIds),
+    };
+  }
+
+  handleIdToggle = (id: string) => {
+    this.setState(state => ({
+      selectedIds: xor(state.selectedIds, [id]),
+      isEverythingSelected: false,
+    }));
+  };
+  handleAllIdsToggle = (select: boolean) => {
+    const {pageIds} = this.props;
+    this.setState({
+      selectedIds: select ? [...pageIds] : [],
+      isEverythingSelected: select,
+    });
+  };
+  handlePageIdsToggle = (select: boolean) => {
+    const {pageIds} = this.props;
+    this.setState(state => ({
+      selectedIds: select
+        ? uniq([...state.selectedIds, ...pageIds])
+        : state.selectedIds.filter(id => !pageIds.includes(id)),
+      isEverythingSelected: false,
+    }));
+  };
+
+  render() {
+    const {pageIds, children, noticeColumns, allIdsCount, maxIdsCount} = this.props;
+    const {selectedIds, isEverythingSelected} = this.state;
+
+    const isPageSelected =
+      pageIds.length > 0 && pageIds.every(id => selectedIds.includes(id));
+
+    const renderProps: RenderProps = {
+      selectedIds,
+      isEverythingSelected,
+      isPageSelected,
+      onIdToggle: this.handleIdToggle,
+      onAllIdsToggle: this.handleAllIdsToggle,
+      onPageIdsToggle: this.handlePageIdsToggle,
+      bulkNotice: (
+        <BulkNotice
+          allRowsCount={allIdsCount}
+          selectedRowsCount={selectedIds.length}
+          onCancelAllRows={() => this.handleAllIdsToggle(false)}
+          onSelectAllRows={() => this.handleAllIdsToggle(true)}
+          columnsCount={noticeColumns}
+          isPageSelected={isPageSelected}
+          isEverythingSelected={isEverythingSelected}
+          bulkLimit={maxIdsCount}
+        />
+      ),
```

Comment:
I will be using this bulkController in an upcoming feature but would like to eventually also replace the one on issues stream.

I made this component with this intention in mind. I believe we should already have all the data exposed for creating customized notice.

I changed the bulkNotice to be a function though. So that it's only called when the implementor wants it.